refact!: Clean up `VersionsButton`

distantnative opened the pull request:
## Description
<!-- 
Add info about why this PR exists and the decisions that went into it.
This info is meant for the reviewer of this PR.
 
You may keep it short or omit it if it's a simple PR. Please add more
context and a summary of changes if it's a more complex PR. 

Make sure to point your PR to the relevant develop branches, e.g.
`develop-patch`, `develop-minor` or `v6/develop`.

How to contribute: https://contribute.getkirby.com
-->


Splitting up the code in `VersionsButton` into smaller methods - makes it easier to customize it a bit further down the road in a v6 PR. But I think these changes could already go into v5.1.

## Changelog
### Breaking change
- `$versionId` parameter for `Kirby\Panel\Ui\Button\VersionsButton::__construct()` has been renamed to `$mode`

lukasbestle commented on src/Panel/Ui/Buttons/VersionsButton.php:
```diff
@@ -21,37 +21,93 @@ class VersionsButton extends ViewButton
 	public function __construct(
 		ModelWithContent $model,
-		VersionId|string $versionId = 'latest'
+		public VersionId|string $versionId = 'latest'
```
Having an instance prop that can carry both types is confusing and error-prone IMO.

`Suggestions` for possible solutions:

- We could create a new `PreviewVersionId` class that supports the `compare` case. However it probably shouldn't `extend VersionId` because of [type-safety](https://www.php.net/manual/en/language.enumerations.object-differences.inheritance.php). So not really clean either.
- We could always convert the passed value to a string in the constructor as we did before. Then the instance prop can always be a string.

distantnative replied:
You are right, but given the use case and rest of the code, I think this should always be a string then (maybe also different name as it already can also be "compare" and could be something like "edit" in the future). I will change it.

distantnative replied:
@lukasbestle Actually, I think we have to leave it like this for 5.1.0 to not introduce that breaking change. We can adapt it directly on `v6/develop` after merging. What do you think?

lukasbestle replied:
Which breaking change? Before this PR, we didn't have an instance prop at all. The constructor argument can stay the same and will be normalized in the constructor.

distantnative replied:
But the constructor parameter existed already before this PR and we would also rename it then, so that would be a breaking change.

lukasbestle replied:
Ah, the renaming. Yes, that is a good point. Then I agree to move the breaking change to a separate PR for v6.

distantnative replied:
Actually, we do not need to have this in v5.1 in general. I moved it to 5.1 as we try to back port as much as possible. But in this case, I think it then doesn't make sense to split this up further but rather do the whole PR for v6.

bastianallgeier replied:
Yes, let's move it to v6

lukasbestle reviewed (commented):
Don't we also need to update `buttons.php` in the site area to use the new `$mode` argument?

distantnative committed:
refact!: Clean up `VersionsButton`

BREAKING CHANGE:  `$versionId` parameter for `Kirby\Panel\Ui\Button\VersionsButton::__construct()` has been renamed to `$mode`

lukasbestle commented on src/Panel/Ui/Button/VersionsButton.php:
```diff
@@ -21,37 +21,94 @@ class VersionsButton extends ViewButton
+
+	/**
+	 * Returns the button icon based on the view's mode
+	 * @since 5.1.0
```
`Todo`: These versions need to be updated to 6.0.0.

lukasbestle commented on src/Panel/Ui/Button/VersionsButton.php:
```diff
@@ -21,37 +21,94 @@ class VersionsButton extends ViewButton
+	 * Returns the button icon based on the view's mode
+	 * @since 5.1.0
+	 */
+	public function icon(): string
```
`Question`: Should this methods that is called in the constructor actually be `public`? Public methods overshadow `Component::__call()`, so `$versionsButton->icon()` for example will always return the dynamic value based on the mode and not the prop value. As the `$icon` property is `public` as well, one could set the property to a different value on the instance, creating a discrepancy between the property and return value of the method.

In this case, this likely won't cause any problems, I'm asking more about the general coding pattern for such component classes.

distantnative replied:
@lukasbestle I think in this case (but also in a looot of other places in Kirby), we use public methods to be able to test them more easily where based on the general usage they could be protected instead. Maybe if we could find an easy/lean unit test style to keep the protected but still test the method in isolation. I roughly know that one could turn a protected method public via reflections in the unit tests, but IMO that has quite a bit of boilerplate code attached to it, which is why I think in many places we have stayed away from it (and rather made the method public).

lukasbestle replied:
Yes, that is definitely true. I have used this pattern as well and generally it has worked quite well.

I think in this case however, we don't need to make the method public to test it. If we make it `protected`, we can still call `$versionsButton->icon()`, but PHP will then automatically fall back to the magic `__call()` method, thus accessing the `$icon` property that we want to test.

distantnative replied:
@lukasbestle I have changed it to protected.

distantnative committed:
fix: Updates from code review
